File: src/augment.cpp
/*
Philip Englund Mathieu
CS5330 Spring 2023
Object database creation program
*/

#include <cstdio>             // lots of standard C/C++ including printf, scanf
#include <cstring>            // functions for working with strings/char arrays
#include <opencv2/opencv.hpp> // OpenCV main include file
#include <chrono>             // for timing

int capdev_init(cv::VideoCapture *&capdev)
{
    capdev = new cv::VideoCapture(0);

    // configurations for my particular webcam
    capdev->set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));
    capdev->set(cv::CAP_PROP_FRAME_WIDTH, 640);  // Setting the width of the video
    capdev->set(cv::CAP_PROP_FRAME_HEIGHT, 480); // Setting the height of the video

    // make sure the capture device is configured correctly
    if (!capdev->isOpened())
    {
        printf("Unable to open video device\n");
        return (-1);
    }

    // get some properties of the image
    cv::Size refS((int)capdev->get(cv::CAP_PROP_FRAME_WIDTH),
                  (int)capdev->get(cv::CAP_PROP_FRAME_HEIGHT));
    printf("Expected size: %d %d\n", refS.width, refS.height);
    return (0);
}
void print_matrix(cv::Mat mat)
{
    for (int i = 0; i < mat.rows; i++)
    {
        printf("[");
        for (int j = 0; j < mat.cols; j++)
        {
            printf(" %3.2f ", mat.at<double>(i, j));
        }
        printf("]\n");
    }
}

void sprint_vector(cv::Mat vect, char *chars)
{
    strcat(chars, "(");
    for (int i = 0; i < vect.rows * vect.cols; i++)
    {
        if (i > 0)
            strcat(chars, ", ");
        char element[16];
        sprintf(element, "%3.3f", vect.at<double>(i));
        strcat(chars, element);
    }
    strcat(chars, ")");
}

void draw_axes(cv::Mat frame, cv::Mat rvec, cv::Mat tvec, cv::Mat cameraMatrix, cv::Mat distCoeffs)
{
    std::vector<cv::Point3f> axes_in = {cv::Point3f(0, 0, 0), cv::Point3f(3, 0, 0), cv::Point3f(0, 3, 0), cv::Point3f(0, 0, 3)};
    std::vector<cv::Point2f> axes;
    cv::projectPoints(axes_in, rvec, tvec, cameraMatrix, distCoeffs, axes);
    cv::line(frame, axes[0], axes[1], cv::Scalar(255, 0, 0), 3);
    cv::line(frame, axes[0], axes[2], cv::Scalar(0, 0, 255), 3);
    cv::line(frame, axes[0], axes[3], cv::Scalar(0, 255, 0), 3);
    cv::putText(frame, "x", axes[1], 0, 1, cv::Scalar(255, 0, 0), 1);
    cv::putText(frame, "y", axes[2], 0, 1, cv::Scalar(0, 255, 0), 1);
    cv::putText(frame, "z", axes[3], 0, 1, cv::Scalar(0, 0, 255), 1);
}

void draw_cube(cv::Mat frame, cv::Mat rvec, cv::Mat tvec, cv::Mat cameraMatrix, cv::Mat distCoeffs)
{
    std::vector<cv::Point3f> points_in;
    for (int i = 0; i <= 3; i += 3)
    {
        for (int j = 0; j <= 3; j += 3)
        {
            for (int k = 0; k <= 3; k += 3)
            {
                points_in.push_back(cv::Point3f(i, j, k + 3));
            }
        }
    }
    std::vector<cv::Point2f> points;
    cv::projectPoints(points_in, rvec, tvec, cameraMatrix, distCoeffs, points);
    for (int i = 0; i < points.size() - 1; i++)
    {
        for (int j = i + 1; j < points.size(); j++)
        {
            cv::line(frame, points[i], points[j], cv::Scalar(255, 0, 0), 1);
        }
    }
}

void draw_sphere(cv::Mat frame, cv::Mat rvec, cv::Mat tvec, cv::Mat cameraMatrix, cv::Mat distCoeffs)
{
    float r = 3;
    std::vector<cv::Point3f> points_in;
    for (float phi = 0; phi < M_PI; phi += 0.1)
    {
        for (float theta = 0; theta < 2 * M_PI; theta += 0.1)
        {
            points_in.push_back(cv::Point3f(r * sin(theta) * cos(phi) + r, r * sin(theta) * sin(phi) + r, r * cos(theta) + r));
        }
    }
    std::vector<cv::Point2f> points;
    cv::projectPoints(points_in, rvec, tvec, cameraMatrix, distCoeffs, points);
    for (int i = 0; i < points.size() - 1; i++)
    {
        cv::line(frame, points[i], points[i + 1], cv::Scalar(255, 0, 255), 1);
    }
}

void draw_torus(cv::Mat frame, cv::Mat rvec, cv::Mat tvec, cv::Mat cameraMatrix, cv::Mat distCoeffs)
{
    float r = 1;
    float r_inner = 3;
    std::vector<cv::Point3f> points_in;
    for (float phi = 0; phi < 2 * M_PI; phi += 0.1)
    {
        for (float theta = 0; theta < 4 * M_PI; theta += 0.1)
        {
            points_in.push_back(cv::Point3f((r_inner + r * sin(theta)) * cos(phi), (r_inner + r * sin(theta)) * sin(phi), r * cos(theta) + r));
        }
    }
    std::vector<cv::Point2f> points;
    cv::projectPoints(points_in, rvec, tvec, cameraMatrix, distCoeffs, points);
    for (int i = 0; i < points.size() - 1; i++)
    {
        cv::line(frame, points[i], points[i + 1], cv::Scalar(0, 255, 0), 5);
    }
}

void draw_knot(cv::Mat frame, cv::Mat rvec, cv::Mat tvec, cv::Mat cameraMatrix, cv::Mat distCoeffs)
{
    float s = 3;
    float p = 3;
    float q = 2;
    std::vector<cv::Point3f> points_in;
    for (float t = 0; t < 2 * M_PI; t += 0.05)
    {
        points_in.push_back(cv::Point3f(s * cos(p * t) * (3 + cos(q * t)), s * sin(p * t) * (3 + cos(q * t)), s * sin(q * t)));
    }
    std::vector<cv::Point2f> points;
    cv::projectPoints(points_in, rvec, tvec, cameraMatrix, distCoeffs, points);
    for (int i = 0; i < points.size() - 1; i++)
    {
        cv::line(frame, points[i], points[i + 1], cv::Scalar(0, 0, 255), 5);
    }
}

void print_vecs(cv::Mat frame, cv::Mat rvec, cv::Mat tvec)
{
    char chars[256] = "Rotation: ";
    sprint_vector(rvec, chars);
    cv::putText(frame, chars, cv::Point(10, frame.rows - 40), 0, 0.5, cv::Scalar(255, 255, 255), 1);
    strcpy(chars, "Translation: ");
    sprint_vector(tvec, chars);
    cv::putText(frame, chars, cv::Point(10, frame.rows - 20), 0, 0.5, cv::Scalar(255, 255, 255), 1);
}

int main(int argc, char *argv[])
{

    // create a window and allocate an image
    cv::namedWindow("Augmented Reality", 1); // identifies a window

    // create the point set for the checkerboard
    std::vector<cv::Vec3f> point_set;
    for (int j = 0; j < 9; j++)
    {
        for (int i = 0; i < 6; i++)
        {
            point_set.push_back(cv::Vec3f(i, -j, 0));
        }
    }

    // load camera calibration
    cv::Mat cameraMatrix, distCoeffs;
    cv::FileStorage fs("./data/calibration.xml", cv::FileStorage::READ);
    fs["cameraMatrix"] >> cameraMatrix;
    fs["distCoeffs"] >> distCoeffs;
    fs.release();

    print_matrix(cameraMatrix);
    print_matrix(distCoeffs);

    enum Object
    {
        SPHERE = 1 << 0,
        CUBE = 1 << 2,
        TORUS = 1 << 3,
        KNOT = 1 << 4
    };
    int current_objects = CUBE;

    int output_count = 0;

    // create and open the video capture device
    cv::VideoCapture *capdev;
    capdev_init(capdev);

    // timer variables
    std::chrono::time_point<std::chrono::system_clock> start, end;
    start = std::chrono::system_clock::now();

    // main loop
    for (;;)
    {
        cv::Mat frame;
        *capdev >> frame; // get a new frame from the camera, treat as a stream
        if (frame.empty())
        {
            printf("frame is empty\n");
            break;
        }

        // find chessboard corners
        cv::Mat viewGray;
        cv::cvtColor(frame, viewGray, cv::COLOR_BGR2GRAY);
        cv::Size patternsize(6, 9);       // interior number of corners
        std::vector<cv::Point2f> corners; // this will be filled by the detected corners
        bool patternfound = cv::findChessboardCorners(viewGray, patternsize, corners, cv::CALIB_CB_ADAPTIVE_THRESH + cv::CALIB_CB_NORMALIZE_IMAGE + cv::CALIB_CB_FAST_CHECK);

        if (patternfound)
        {
            cv::cornerSubPix(viewGray, corners, cv::Size(11, 11), cv::Size(-1, -1),
                             cv::TermCriteria(cv::TermCriteria::EPS + cv::TermCriteria::COUNT, 30, 0.1));
            cv::Mat rvec, tvec;
            cv::solvePnP(point_set, corners, cameraMatrix, distCoeffs, rvec, tvec);

            draw_axes(frame, rvec, tvec, cameraMatrix, distCoeffs);
            print_vecs(frame, rvec, tvec);
            if (current_objects & CUBE)
                draw_cube(frame, rvec, tvec, cameraMatrix, distCoeffs);
            if (current_objects & SPHERE)
                draw_sphere(frame, rvec, tvec, cameraMatrix, distCoeffs);
            if (current_objects & TORUS)
                draw_torus(frame, rvec, tvec, cameraMatrix, distCoeffs);
            if (current_objects & KNOT)
                draw_knot(frame, rvec, tvec, cameraMatrix, distCoeffs);
        }

        // show FPS on window
        end = std::chrono::system_clock::now();
        std::chrono::duration<double> elapsed_seconds = end - start;
        double seconds = std::chrono::duration<double>(elapsed_seconds).count();
        int fps = 1 / seconds;
        cv::putText(frame, std::to_string(fps), cv::Point2f(7, 70), cv::FONT_HERSHEY_SIMPLEX, 2, cv::Scalar(100, 255, 0), 2, cv::LINE_AA);

        cv::imshow("Augmented Reality", frame);
        start = std::chrono::system_clock::now();

        // see if there is a waiting keystroke
        char key = cv::waitKey(10);

        if (key == 'q')
        {
            break;
        }

        switch (key)
        {
        case 'c':
            current_objects ^= CUBE;
            break;
        case 'o':
            current_objects ^= SPHERE;
            break;
        case 't':
            current_objects ^= TORUS;
            break;
        case 'k':
            current_objects ^= KNOT;
            break;
        case 's':
            char filename[128];
            sprintf(filename, "./data/ar_%d.jpg", output_count);
            output_count++;
            cv::imwrite(filename, frame);
            break;
        }
    }
}

File: src/augment.py
'''
Philip Englund Mathieu
CS5330 Spring 2023
Augmented Reality Program (Python Implementation)
'''

import cv2 as cv                    # OpenCV
import numpy as np                  # matrix library
import xml.etree.ElementTree as ET  # for parsing xml
from math import *                  # constants and basic functions
import time                         # for timing performance

# global constants
PATTERN_SIZE = (9,6)
CRITERIA = (cv.TERM_CRITERIA_EPS + cv.TERM_CRITERIA_COUNT, 30, 0.1)

def load_from_opencv_xml(filename, elementname, dtype='float32'):
    '''
    Loads camera matrix and distortion coefficients from an OpenCV .xml file.
    Inputs: filename, name of element to retrieve, datatype of element
    Outputs: element
    This function is copied from StackExchange: https://stackoverflow.com/questions/11025477/error-loading-opencv-xml-file-with-python
    '''
    try:
        tree = ET.parse(filename)
        rows = int(tree.find(elementname).find('rows').text)
        cols = int(tree.find(elementname).find('cols').text)
        return np.fromstring(tree.find(elementname).find('data').text, dtype, count=rows*cols, sep=' ').reshape((rows, cols))
    except Exception as e: 
        print(e)
        return None

def draw_axes(frame, rvec, tvec, cameraMatrix, distCoeffs):
    '''
    This function draws XYZ axes on an image.
    Inputs: image to draw on, rotation vector, translation vector,
        camera matrix, distortion coefficients
    Outputs: None
    '''
    axes = np.array([[0, 0, 0], [3, 0, 0], [0, 3, 0], [0, 0, 3]], dtype=np.float32)
    axes, _ = cv.projectPoints(axes, rvec, tvec, cameraMatrix, distCoeffs)
    axes = axes.astype(np.int32).reshape((-1,2))
    cv.line(frame, axes[0], axes[1], (255, 0, 0), 3)
    cv.line(frame, axes[0], axes[2], (0, 0, 255), 3)
    cv.line(frame, axes[0], axes[3], (0, 255, 0), 3)
    cv.putText(frame, "x", axes[1], 0, 1, (255, 0, 0), 1)
    cv.putText(frame, "y", axes[2], 0, 1, (0, 255, 0), 1)
    cv.putText(frame, "z", axes[3], 0, 1, (0, 0, 255), 1)

def draw_cube(frame, rvec, tvec, cameraMatrix, distCoeffs):
    '''
    This function draws a cube on an image.
    Inputs: image to draw on, rotation vector, translation vector,
        camera matrix, distortion coefficients
    Outputs: None
    '''
    points = np.array([[i, j, k] for i in range(3) for j in range(3) for k in range(3)], dtype=np.float32)
    points, _ = cv.projectPoints(points, rvec, tvec, cameraMatrix, distCoeffs)
    points = points.astype(np.int32).reshape((-1,2))
    [cv.line(frame, points[i], points[j], (255, 0, 0), 1) for i in range(0, points.shape[0]-1) for j in range(1, points.shape[0])]

def draw_sphere(frame, rvec, tvec, cameraMatrix, distCoeffs):
    '''
    This function draws a sphere on an image.
    Inputs: image to draw on, rotation vector, translation vector,
        camera matrix, distortion coefficients
    Outputs: None
    '''
    r = 3
    points = [(r * sin(theta) * cos(phi) + r, r * sin(theta) * sin(phi) + r, r * cos(theta) + r) for phi in np.linspace(0, pi, 10) for theta in np.linspace(0, 2*pi, 10) ]
    points = np.array(points, dtype=np.float32)
    points, _ = cv.projectPoints(points, rvec, tvec, cameraMatrix, distCoeffs)
    points = points.astype(np.int32).reshape((-1,2))
    [cv.line(frame, points[i], points[i + 1], (255, 0, 255), 1) for i in range(0, points.shape[0]-1)]

def draw_torus(frame, rvec, tvec, cameraMatrix, distCoeffs):
    '''
    This function draws a torus on an image.
    Inputs: image to draw on, rotation vector, translation vector,
        camera matrix, distortion coefficients
    Outputs: None
    '''
    r = 1
    r_inner = 3
    points = [((r_inner + r * sin(theta)) * cos(phi), (r_inner + r * sin(theta)) * sin(phi), r * cos(theta) + r) for phi in np.linspace(0, 2*pi, 10) for theta in np.linspace(0, 4*pi, 21)]
    points = np.array(points, dtype=np.float32)
    points, _ = cv.projectPoints(points, rvec, tvec, cameraMatrix, distCoeffs)
    points = points.astype(np.int32).reshape((-1,2))
    [cv.line(frame, points[i], points[i + 1], (0, 255, 0), 5) for i in range(0, points.shape[0]-1)]

def draw_knot(frame, rvec, tvec, cameraMatrix, distCoeffs):
    '''
    This function draws a knot on an image.
    Inputs: image to draw on, rotation vector, translation vector,
        camera matrix, distortion coefficients
    Outputs: None
    '''
    s = 3
    p = 3
    q = 2
    points = [(s * cos(p * t) * (3 + cos(q * t)), s * sin(p * t) * (3 + cos(q * t)), s * sin(q * t)) for t in np.linspace(0, 2*pi, 100)]
    points = np.array(points, dtype=np.float32)
    points, _ = cv.projectPoints(points, rvec, tvec, cameraMatrix, distCoeffs)
    points = points.astype(np.int32).reshape((-1,2))
    [cv.line(frame, points[i], points[i + 1], (0, 0, 255), 10) for i in range(0, points.shape[0]-1)]

if __name__ == "__main__":

    # create a window and allocate an image
    cv.namedWindow("Augmented Reality", 1); # identifies a window

    # create the point set for the checkerboard
    objp = np.array([[i, j, 0] for i in range(PATTERN_SIZE[1]) for j in range(PATTERN_SIZE[0])], dtype=np.float32)

    # load camera calibration
    cameraMatrix = load_from_opencv_xml("./data/calibration.xml", "cameraMatrix", np.double)
    distCoeffs = load_from_opencv_xml("./data/calibration.xml", "distCoeffs", np.double)
    print(cameraMatrix)
    print(distCoeffs)

    current_objects =  {
        "sphere": False,
        "cube": True,
        "torus": False,
        "knot": False
    }

    output_count = 0

    cap = cv.VideoCapture(0)
    # configurations for my particular webcam
    cap.set(cv.CAP_PROP_FOURCC, cv.VideoWriter_fourcc(*'MJPG'))
    cap.set(cv.CAP_PROP_FRAME_WIDTH, 640)
    cap.set(cv.CAP_PROP_FRAME_HEIGHT, 480)

    # timing functions
    prev_frame_time = 0
    new_frame_time = 0

    # main loop
    while (True):
        # Capture frame-by-frame
        ret, frame = cap.read()
        # if frame is read correctly ret is True
        if not ret:
            print("Can't receive frame (stream end?). Exiting ...")
            break

        # find chessboard corners
        gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        # Find the chess board corners
        ret, corners = cv.findChessboardCorners(gray, PATTERN_SIZE, None)

        if ret: # corners successfully found
            corners = cv.cornerSubPix(gray, corners, (11, 11), (-1, -1), CRITERIA)
            ret, rvec, tvec = cv.solvePnP(objp, corners, cameraMatrix, distCoeffs)

            # draw the pattern if successful
            if ret:
                 cv.drawChessboardCorners(frame, PATTERN_SIZE, corners, ret)
            
            # add the other drawings to the image
            draw_axes(frame, rvec, tvec, cameraMatrix, distCoeffs)
            if (current_objects["cube"]):
                draw_cube(frame, rvec, tvec, cameraMatrix, distCoeffs)
            if (current_objects["sphere"]):
                draw_sphere(frame, rvec, tvec, cameraMatrix, distCoeffs)
            if (current_objects["torus"]):
                draw_torus(frame, rvec, tvec, cameraMatrix, distCoeffs)
            if (current_objects["knot"]):
                draw_knot(frame, rvec, tvec, cameraMatrix, distCoeffs)
        
        # show FPS on window
        new_frame_time = time.time()
        fps = str(int(1/(new_frame_time-prev_frame_time)))
        cv.putText(frame, fps, (7, 70), cv.FONT_HERSHEY_SIMPLEX, 2, (100, 255, 0), 2, cv.LINE_AA)

        cv.imshow("Augmented Reality", frame)
        prev_frame_time = time.time()

        # see if there is a waiting keystroke
        key = cv.waitKey(10)

        # parse key presses
        if key == ord('q'):
            break
        elif key == ord('c'):
            current_objects["cube"] = not current_objects["cube"]
        elif key == ord('o'):
            current_objects["sphere"] = not current_objects["sphere"]
        elif key == ord('k'):
            current_objects["knot"] = not current_objects["knot"]
        elif key == ord('t'):
            current_objects["torus"] = not current_objects["torus"]
        elif key == ord('s'):
            output_count += 1
            cv.imwrite("./data/ar_py_{:d}.jpg".format(output_count), frame)

    cv.destroyAllWindows()
    cap.release()